types/data_source.go

```diff
@@ -0,0 +1,131 @@
+package types
+
+import (
+	"fmt"
+	"github.com/pkg/errors"
+	"strings"
+)
+
+const (
+	SnapPrefix = "snap"
+	VolPrefix  = "vol"
+)
+
+func NewVolumeDataSource(volumeDataSourceType string, parameters ...string) (dataSource VolumeDataSource, err error) {
+	defer func() {
+		err = errors.Wrapf(err, "cannot create new VolumeDataSource")
+	}()
+
+	switch volumeDataSourceType {
+	case VolumeDataSourceTypeVolume:
+		if len(parameters) != 1 {
+			return VolumeDataSource(""), fmt.Errorf("VolumeDataSourceTypeVolume requires 1 parameter but %v were given", len(parameters))
+		}
+		volumeName := parameters[0]
+		dataSource := VolumeDataSource(fmt.Sprintf("%v://%s", VolPrefix, volumeName))
+		return dataSource, nil
+	case VolumeDataSourceTypeSnapshot:
+		if len(parameters) != 2 {
+			return VolumeDataSource(""), fmt.Errorf("VolumeDataSourceTypeSnapshot requires 2 parameters but %v were given", len(parameters))
+		}
+		volumeName := parameters[0]
+		snapshotName := parameters[1]
+		dataSource := VolumeDataSource(fmt.Sprintf("%v://%s/%s", SnapPrefix, volumeName, snapshotName))
+		return dataSource, nil
+	default:
+		return VolumeDataSource(""), fmt.Errorf("invalid volumeDataSourceType %v", volumeDataSourceType)
+	}
+}
+
+func IsValidVolumeDataSource(vds VolumeDataSource) bool {
+	split := strings.Split(string(vds), "://")
+	if len(split) != 2 {
+		return false
+	}
+	prefix := split[0]
+	suffix := split[1]
```

Comment:
To avoid duplicate, you can directly use `decodeDataSourceTypeSnapshot` or `decodeDataSourceTypeVolume` based on `prefix` value

Reply:
Thank you. Fixed